metaseq/modules/multihead_attention.py

```diff
@@ -312,6 +314,24 @@ def forward(
             # In this branch incremental_state is never None
             assert incremental_state is not None
             incremental_state = self._set_input_buffer(incremental_state, saved_state)
+
+        if isinstance(q, ShardedTensor):
+            sharding_spec = q.sharding_spec()
+            sharding_spec.dim = 0
+            st_size = (bsz * self.num_heads, q.size(1), self.head_dim)
+            q = ShardedTensor._init_from_local_tensor(
+                q.local_tensor(), sharding_spec, st_size
+            )
+            if k is not None:
+                st_size = (bsz * self.num_heads, k.size(1), self.head_dim)
+                k = ShardedTensor._init_from_local_tensor(
+                    k.local_tensor(), sharding_spec, st_size
+                )
+            if v is not None:
+                st_size = (bsz * self.num_heads, v.size(1), self.head_dim)
+                v = ShardedTensor._init_from_local_tensor(
```

Comment:
Why do we need to create ShardedTensors here? Doesn't `shard_module` appropriately shard multihead attention?

Reply:
Megatron implicitly change the sharding spec during view.